File: src/point.cpp
#include "point.h"

Point::Point(double x_, double y_) : x(x_), y(y_) {}

double Point::cross_z(const Point& other) const {
  return x * other.y - y * other.x;
}

bool Point::is_right_of(const Point& other) const {
  if (x == other.x)
    return y > other.y;
  else
    return x > other.x;
}

bool Point::operator==(const Point& other) const {
  return x == other.x && y == other.y;
}

bool Point::operator!=(const Point& other) const {
  return x != other.x || y != other.y;
}

Point Point::operator*(const double& multiplier) const {
  return Point(x * multiplier, y * multiplier);
}

const Point& Point::operator+=(const Point& other) {
  x += other.x;
  y += other.y;
  return *this;
}

const Point& Point::operator-=(const Point& other) {
  x -= other.x;
  y -= other.y;
  return *this;
}

Point Point::operator+(const Point& other) const {
  return Point(x + other.x, y + other.y);
}

Point Point::operator-(const Point& other) const {
  return Point(x - other.x, y - other.y);
}

std::ostream& operator<<(std::ostream& os, const Point& xy) {
  return os << '(' << xy.x << ' ' << xy.y << ')';
}

File: seidel/point.py
from reprit.base import generate_repr

from .hints import Coordinate


class Point:
    __slots__ = 'x', 'y'

    def __init__(self, x: Coordinate, y: Coordinate) -> None:
        self.x = x
        self.y = y

    __repr__ = generate_repr(__init__)

    def __add__(self, other: 'Point') -> 'Point':
        return Point(self.x + other.x, self.y + other.y)

    def __eq__(self, other: 'Point') -> bool:
        return (self.x == other.x and self.y == other.y
                if isinstance(other, Point)
                else NotImplemented)

    def __mul__(self, scale: Coordinate) -> 'Point':
        return Point(self.x * scale, self.y * scale)

    def __sub__(self, other: 'Point') -> 'Point':
        return Point(self.x - other.x, self.y - other.y)

    def cross_z(self, other: 'Point') -> Coordinate:
        return self.x * other.y - self.y * other.x

    def is_right_of(self, other: 'Point') -> bool:
        return (self.x, self.y) > (other.x, other.y)
